samcli/commands/package/s3_uploader.py

```diff
@@ -0,0 +1,210 @@
+        """
+
+        # This construction of remote_path is critical to preventing duplicate
+        # uploads of same object. Uploader will check if the file exists in S3
+        # and re-upload only if necessary. So the template points to same file
+        # in multiple places, this will upload only once
+
+        filemd5 = self.file_checksum(file_name)
+        remote_path = filemd5
+        if extension:
+            remote_path = remote_path + "." + extension
+
+        return self.upload(file_name, remote_path)
+
+    def file_exists(self, remote_path):
+        """
+        Check if the file we are trying to upload already exists in S3
+
+        :param remote_path:
+        :return: True, if file exists. False, otherwise
+        """
+
+        try:
+            # Find the object that matches this ETag
+            self.s3.head_object(Bucket=self.bucket_name, Key=remote_path)
+            return True
+        except botocore.exceptions.ClientError:
+            # Either File does not exist or we are unable to get
+            # this information.
+            return False
+
+    def make_url(self, obj_path):
+        return "s3://{0}/{1}".format(self.bucket_name, obj_path)
+
+    def file_checksum(self, file_name):
+
+        with open(file_name, "rb") as file_handle:
+            md5 = hashlib.md5()
+            # Read file in chunks of 4096 bytes
+            block_size = 4096
+
+            # Save current cursor position and reset cursor to start of file
+            curpos = file_handle.tell()
+            file_handle.seek(0)
+
+            buf = file_handle.read(block_size)
+            while buf:
+                md5.update(buf)
+                buf = file_handle.read(block_size)
+
+            # Restore file cursor's position
+            file_handle.seek(curpos)
+
+            return md5.hexdigest()
+
+    def to_path_style_s3_url(self, key, version=None):
+        """
+            This link describes the format of Path Style URLs
+            http://docs.aws.amazon.com/AmazonS3/latest/dev/UsingBucket.html#access-bucket-intro
+        """
+        base = self.s3.meta.endpoint_url
+        result = "{0}/{1}/{2}".format(base, self.bucket_name, key)
+        if version:
+            result = "{0}?versionId={1}".format(result, version)
+
+        return result
+
+
+class ProgressPercentage:
+    # This class was copied directly from S3Transfer docs
+
+    def __init__(self, filename, remote_path):
+        self._filename = filename
+        self._remote_path = remote_path
+        self._size = float(os.path.getsize(filename))
+        self._seen_so_far = 0
+        self._lock = threading.Lock()
+
+    def on_progress(self, bytes_transferred, **kwargs):
+
+        # To simplify we'll assume this is hooked up
+        # to a single filename.
+        with self._lock:
+            self._seen_so_far += bytes_transferred
+            percentage = (self._seen_so_far / self._size) * 100
+            sys.stdout.write(
```

Comment:
Should this be stdout or stderr?
https://github.com/awslabs/aws-sam-cli/issues/970

Reply:
Good thought, will change to stderr